src/main/kotlin/com/exactpro/th2/ws/client/Main.kt

```diff
@@ -0,0 +1,180 @@
+
+    val factory = args.runCatching(CommonFactory::createFromArguments).getOrElse {
+        LOGGER.error(it) { "Failed to create common factory with arguments: ${args.joinToString(" ")}" }
+        CommonFactory()
+    }.apply { resources += "factory" to ::close }
+
+    val mapper = JsonMapper.builder()
+        .addModule(KotlinModule(nullIsSameAsDefault = true))
+        .addModule(SimpleModule().addAbstractTypeMapping(IHandlerSettings::class.java, handlerSettingType))
+        .build()
+
+    val settings = factory.getCustomConfiguration(Settings::class.java, mapper)
+    val eventRouter = factory.eventBatchRouter
+    val messageRouter = factory.messageRouterMessageGroupBatch
+
+    run(settings, eventRouter, messageRouter, handler) { resource, destructor ->
+        resources += resource to destructor
+    }
+} catch (e: Exception) {
+    LOGGER.error(e) { "Uncaught exception. Shutting down" }
+    exitProcess(1)
+}
+
+fun run(
+    settings: Settings,
+    eventRouter: MessageRouter<EventBatch>,
+    messageRouter: MessageRouter<MessageGroupBatch>,
+    handler: IHandler,
+    registerResource: (name: String, destructor: () -> Unit) -> Unit
+) {
+    val connectionId = ConnectionID.newBuilder().setSessionAlias(settings.sessionAlias).build()
+
+    val rootEventId = eventRouter.storeEvent(Event.start().apply {
+        endTimestamp()
+        name("WS client '${settings.sessionAlias}' ${Instant.now()}")
+        type("Microservice")
+    }).id
+
+    val generateSequence = Instant.now().run {
+        AtomicLong(epochSecond * SECONDS.toNanos(1) + nano)
+    }::incrementAndGet
+
+    settings.handlerSettings.runCatching(handler::init).onFailure {
+        LOGGER.error(it) { "Failed to init request handler" }
+        eventRouter.storeEvent(rootEventId, "Failed to init request handler", "Error", it)
+        throw it
+    }
+
+    val onMessage = { message: ByteArray, _: Boolean, direction: Direction ->
+        messageRouter.send(message.toBatch(connectionId, direction, generateSequence()))
+    }
+
+    val onEvent = { cause: Throwable?, message: () -> String ->
+        val type = if(cause != null) "Error" else "Info"
+        eventRouter.storeEvent(rootEventId, message(), type, cause)
+    }
+
+    val client = WebSocketClient(
+        URI(settings.uri),
+        handler,
+        onMessage,
+        onEvent
+    ).apply { registerResource("client", ::close) }
+
+    val listener = MessageListener<MessageGroupBatch> { _, message ->
+        message.groupsList.forEach { group ->
+            group.runCatching {
+                require(messagesCount == 1) { "Message group contains more than 1 message" }
+                val message = messagesList[0]
+                require(message.hasRawMessage()) { "Message in the group is not a raw message" }
+                val bytes = message.rawMessage.body.toByteArray()
+                val textual = Utf8.isWellFormed(bytes) //TODO: we should probably get data type from message metadata
```

Comment:
Would we use different ways to publish in one connection? We can add the mode publish option to conn configuration if no.

Reply:
I won't discard the possibility of sending different types of frames in a single connection, but we probably won't need it at first

Reply:
Let's add an enum parameter with strict modes into the configuration. We can extend this enum in the future if it will be required.